src/testrender/cuda/optix_raytracer.cu

```diff
@@ -34,6 +46,172 @@ __device__ __constant__ RenderParams render_params;
+        // TODO: should probably never get here ...
+        return;
+    }
+
+    // Pack the "closure pool" into one of the ShaderGlobals pointers
+    *(int*)&closure_pool[0] = 0;
+    sg.renderstate          = &closure_pool[0];
+
+    // Pack the pointers to the options structs in a faux "context",
+    // which is a rough stand-in for the host ShadingContext.
+    ShadingContextCUDA shading_context;
+    sg.context = &shading_context;
+
+    // Run the OSL callable
+    void* interactive_ptr = reinterpret_cast<void**>(
+        render_params.interactive_params)[shader_id];
+    const unsigned int shaderIdx = shader_id + 0u;
+    optixDirectCall<void, ShaderGlobalsType*, void*, void*, void*, int, void*>(
+        shaderIdx, &sg /*shaderglobals_ptr*/, nullptr /*groupdata_ptr*/,
+        nullptr /*userdata_base_ptr*/, nullptr /*output_base_ptr*/,
+        0 /*shadeindex - unused*/, interactive_ptr /*interactive_params_ptr*/
+    );
+}
+
+
+static inline __device__ void
+trace_ray(OptixTraversableHandle handle, Payload& payload, const float3& origin,
+          const float3& direction, const float tmin)
+{
+    uint32_t p0 = static_cast<uint32_t>(payload.hit_id);
+    uint32_t p1 = __float_as_uint(payload.hit_t);
+    uint32_t p2 = __float_as_uint(payload.hit_u);
+    uint32_t p3 = __float_as_uint(payload.hit_v);
+    optixTrace(handle,                         // handle
+               origin,                         // origin
+               direction,                      // direction
+               tmin,                           // tmin
+               1e13f,                          // tmax
+               0,                              // ray time
+               OptixVisibilityMask(1),         // visibility mask
+               OPTIX_RAY_FLAG_DISABLE_ANYHIT,  // ray flags
+               0,                              // SBT offset
+               1,                              // SBT stride
+               0,                              // miss SBT offset
+               p0, p1, p2, p3);
+    payload.hit_id = static_cast<int32_t>(p0);
+    payload.hit_t  = __uint_as_float(p1);
+    payload.hit_u  = __uint_as_float(p2);
+    payload.hit_v  = __uint_as_float(p3);
+};
+
+
+Intersection
+Scene::intersect(const Ray& r, const float tmax, const unsigned skipID1,
+                 const unsigned /*skipID2*/) const
+{
+    // Trace the ray against the scene. If the ID for the hit matches skipID1,
+    // "nudge" the ray forward by adjusting tmin to exclude the hit interval
+    // and try again.
+    const float epsilon    = 1e-6f;
+    const int num_attempts = 2;
+    float tmin             = epsilon;
+    for (int attempt = 0; attempt < num_attempts; ++attempt) {
+        Payload payload;
+        payload.hit_id = ~0u;
+        trace_ray(handle, payload, V3_TO_F3(r.origin), V3_TO_F3(r.direction),
+                  tmin);
+        if (payload.hit_id == skipID1) {
+            tmin = payload.hit_t + 2.0f * epsilon;
```

Comment:
Could nudge by bumping the integer representation instead. This would let you use a smaller epsilon.

Reply:
That is a trick I was not aware of, and it appears to work. Nifty.